### PR TITLE
Data transfer status bars

### PR DESCRIPTION
**Pull request by chStaiger**
- Switched the transfers from upload/download to _obj_put/_obj_get
- Added status bars in percent according to the tranferred file sizes
- Some small bug fixes

### CODE REVIEW

**Comment by qubixes** [comment]
I tried the upload function for a folder, but it does not seem to work. The problem seems that the collections are not made (in time).

**Comment by qubixes** [comment]
I also had iBridges lock up during connection (first time went okay, but there was nothing in the password field). Second time it locked up after clicking connect. I don't know if it's related to this PR of course.

**Comment by qubixes** [comment]
Another thing I noticed: As far as I understand, the bar denotes the number of bytes, right? Perhaps it would be useful to somewhere put the amount transfered/left/total?

**Comment by chStaiger** [comment]
> I tried the upload function for a folder, but it does not seem to work. The problem seems that the collections are not made (in time).

Oh dear a bug ;). Yes, the folders were not created at all when there was just one folder. Fixed.

**Comment by chStaiger** [comment]
> Another thing I noticed: As far as I understand, the bar denotes the number of bytes, right? Perhaps it would be useful to somewhere put the amount transfered/left/total?

The bar denotes the percentage of Bytes that are transferred. It ranges from 1% to 100%. I can put the percent in there.

**Comment by chStaiger** [comment]
> I also had iBridges lock up during connection (first time went okay, but there was nothing in the password field). Second time it locked up after clicking connect. I don't know if it's related to this PR of course.

I am not sure what you mean with "lock up"? Did it not connect?

File: ibridgesgui/popup_widgets.py [hunk 1]
```diff
@@ -435,38 +451,38 @@ def close_window(self):
...
-
     def irods_path_tree(self):
         """Expand the irods_path if it is a collection."""
         if self.irods_path.collection_exists():
             return "\n".join(
-                    [coll.name for coll in self.irods_path.collection.subcollections]\
-                    + [obj.name for obj in self.irods_path.collection.data_objects ])
+                [coll.name for coll in self.irods_path.collection.subcollections]
+                + [obj.name for obj in self.irods_path.collection.data_objects]
+            )
```
Comment (qubixes) [inline ibridgesgui/popup_widgets.py:476]: I would split this up into several commands.

File: ibridgesgui/popup_widgets.py [hunk 2]
```diff
@@ -481,12 +497,12 @@ def _get_download_params(self):
 
         if not local_path.is_dir():
             self.error_label.setText(
-                    f"Dowload folder {local_path} dows not exist or is not a folder.")
+                f"Dowload folder {local_path} dows not exist or is not a folder."
```
Comment (qubixes) [inline ibridgesgui/popup_widgets.py:500]:
```suggestion
                f"Dowload folder {local_path} does not exist or is not a folder."
```

File: ibridgesgui/popup_widgets.py [hunk 3]
```diff
@@ -527,18 +549,21 @@ def _start_download(self, local_path):
         self.download_thread.start()
 
     def _finish_download(self):
-        self._enable_buttons(True)
+        #self._enable_buttons(True)
```
Comment (qubixes) [inline ibridgesgui/popup_widgets.py:552]: Remove?

File: ibridgesgui/popup_widgets.py [hunk 4]
```diff
@@ -527,18 +549,21 @@ def _start_download(self, local_path):
...
+        self.error_label.setText(text)
 
     def _download_end(self, thread_output: dict):
         self.active_download = False
         if thread_output["error"] == "":
             self.error_label.setText("Download finished.")
         else:
             self.error_label.setText("Errors occurred during download. Consult the logs.")
-        self._enable_buttons(True)
+        #self._enable_buttons(True)
```
Comment (qubixes) [inline ibridgesgui/popup_widgets.py:568]: Remove?

File: ibridgesgui/sync.py [hunk 5]
```diff
@@ -232,7 +233,7 @@ def _start_sync_diff(self, source, target):
             self.sync_diff_thread = SyncThread(env_path, self.logger, source, target, dry_run=True)
         except Exception:
             self.error_label.setText(
-                "Could not instantiate a new session from{env_path}.Check configuration."
+                f"Could not instantiate a new session from{env_path}.Check configuration."
```
Comment (qubixes) [inline ibridgesgui/sync.py:236]:
```suggestion
                f"Could not instantiate a new session from {env_path}. Check configuration."
```

**Review by qubixes — CHANGES_REQUESTED** [review]
Nice improvements! Some small comments/details that I found.

File: ibridgesgui/gui_utils.py [hunk 6]
```diff
@@ -86,10 +86,10 @@ def prep_session_for_copy(session, error_label) -> pathlib.Path:
...
     """Combine the operations of several upload or download dry-runs."""
     ops = operations[0]
-    ops.create_dir = set().union(*[o.create_dir for o in operations])
-    ops.create_collection = set().union(*[o.create_collection for o in operations])
-    _ = [ops.download.extend(o.download) for o in operations]
-    _ = [ops.upload.extend(o.upload) for o in operations]
+    ops.create_dir = set().union(*[o.create_dir for o in operations[1:]])
+    ops.create_collection = set().union(*[o.create_collection for o in operations[1:]])
+    _ = [ops.download.extend(o.download) for o in operations[1:]]
+    _ = [ops.upload.extend(o.upload) for o in operations[1:]]
```
[the remark this answers is not in the document]
Comment (chStaiger) [reply ibridgesgui/gui_utils.py]: Yes, agreed!

File: ibridgesgui/popup_widgets.py [hunk 7]
```diff
@@ -373,20 +384,23 @@ def _start_upload(self, lpaths):
...
+        self.error_label.setText(text)
 
     def _upload_end(self, thread_output: dict):
         self.active_upload = False
         if thread_output["error"] == "":
             self.error_label.setText("Upload finished.")
         else:
             self.error_label.setText("Errors occurred during upload. Consult the logs.")
-        self._enable_buttons(True)
+        #self._enable_buttons(True)
```
[the remark this answers is not in the document]
Comment (chStaiger) [reply ibridgesgui/popup_widgets.py]: Thank for the remark. Yes indeed the naming here is very confusing.

The QThread has by default a slot `finish`. The function `_finish_(search/download/upload...)` connects to that slot and executes the clean up and re-enabling the buttons.
The `upload_end` connects to a slot that I created to retrieve the results from a Qthread and should therefore be called `fetch_result`. I also cleaned up the functionality, i.e. solely fetching results nd for the `_finish_*` solely re-enabling the main GUI.

I hope this is clearer now.

File: ibridgesgui/popup_widgets.py (hunk 1 above)
Comment (chStaiger) [reply ibridgesgui/popup_widgets.py:474]: I will keep it in mind. But for now, with all the other changes I would rather not like to change it yet. Next round?

**Pull request merged**